### PR TITLE
Feature/core data

### PR DESCRIPTION
**Pull request by MaksimK193**

### CODE REVIEW

File: SwiftUIStudy/Screens/CoreData/CoreDataView.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import SwiftUI
+
+struct CoreDataView: View {
+    @StateObject var vm = CoreDataViewModel()
+    
+    var body: some View {
+        List(vm.departments) { department in
+            if let employees = department.employees?.allObjects as? [EmployeeEntity] {
+                NavigationLink(destination: EmployeesView(vm: vm, department: department.name ?? "", employees: employees)) {
+                    Text(department.name ?? "Unknown")
```
Comment (IslamRustamov98) [inline SwiftUIStudy/Screens/CoreData/CoreDataView.swift:18]: maybe we could try to mark "name" as not optional?

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    @Published var departments: [DepartmentEntity] = []
+    @Published var employees: [EmployeeEntity] = []
+    
+    init() {
+        getDepartments()
+        getEmployees()
+    }
+    
+    func getDepartments() {
+        let request = NSFetchRequest<DepartmentEntity>(entityName: "DepartmentEntity")
```
Comment (IslamRustamov98) [inline SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:22]: just FYI, using magic strings is not a good thing, so would be better to have enum with all of our entity names

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    func reset(entity: String) {
+        let deleteFetch = NSFetchRequest<NSFetchRequestResult>(entityName: entity)
+        let deleteRequest = NSBatchDeleteRequest(fetchRequest: deleteFetch)
+        do
+        {
+            try coreDataManager.context.execute(deleteRequest)
+            try coreDataManager.context.save()
+        }
+        catch
+        {
```
Comment (IslamRustamov98) [inline SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:48]: and here

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    
+    func addEmployee(name: String, id: UUID? = UUID(), departmentName: String) {
+        let newEmployee = EmployeeEntity(context: coreDataManager.context)
+        newEmployee.id = id
+        newEmployee.name = name
+        departments.map {
+            if $0.name == departmentName {
+                $0.addToEmployees(newEmployee)
+            }
+        }
```
Comment (IslamRustamov98) [inline SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:68]: Using map for array when you are not doing anything with a returned array is a not nice approach. In this cases for is being used.

But in this particular case, you are iterating over array (in worst case N iterations), don't you think that using that using something like:
if let i = array.firstIndex(where: { $0.name == "Foo" }) {
    return array[i]
}
is going to be better?

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    
+    func delete(_ object: NSManagedObject) {
+        coreDataManager.delete(object)
+        save()
+    }
+    
+    func save() {
+        departments.removeAll()
+        employees.removeAll()
```
Comment (IslamRustamov98) [inline SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:79]: not really sure if it's the best approach, but alright, let's leave it like this

File: SwiftUIStudy/Screens/CoreData/CoreDataView.swift (hunk 1 above)
Comment (MaksimK193) [reply SwiftUIStudy/Screens/CoreData/CoreDataView.swift:18]: It is CoreData generated data model
https://www.atomicbird.com/blog/clash-of-the-optionals/

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        } catch let error {
+            print("Error fetching departments: \(error.localizedDescription)")
+        }
+    }
+    
+    func reset(entity: String) {
+        let deleteFetch = NSFetchRequest<NSFetchRequestResult>(entityName: entity)
+        let deleteRequest = NSBatchDeleteRequest(fetchRequest: deleteFetch)
+        do
+        {
```
[the remark this answers is not in the document]
Comment (MaksimK193) [reply SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:43]: fixed

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift (hunk 3 above)
Comment (MaksimK193) [reply SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:48]: fixed

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift (hunk 4 above)
Comment (MaksimK193) [reply SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:68]: fixed

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift (hunk 5 above)
Comment (MaksimK193) [reply SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift:79]: I don't think it is necessary, but in many guides it was good advice

File: SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift (hunk 2 above)
Comment (MaksimK193) [reply SwiftUIStudy/Screens/CoreData/CoreDataViewModel.swift]: fixed

**Pull request merged**